packages/vite/src/node/plugins/resolve.ts

```diff
@@ -410,59 +422,69 @@ export function resolvePlugin(
-              options.noExternal === true &&
-              // if both noExternal and external are true, noExternal will take the higher priority and bundle it.
-              // only if the id is explicitly listed in external, we will externalize it and skip this error.
-              (options.external === true || !options.external.includes(id))
-            ) {
-              let message = `Cannot bundle Node.js built-in "${id}"`
-              if (importer) {
-                message += ` imported from "${path.relative(
-                  process.cwd(),
-                  importer,
-                )}"`
-              }
-              message += `. Consider disabling environments.${this.environment.name}.noExternal or remove the built-in dependency.`
-              this.error(message)
+        // built-ins
+        // externalize if building for a server environment, otherwise redirect to an empty module
+        if (
+          currentEnvironmentOptions.consumer === 'server' &&
+          isBuiltin(this.environment.config.resolve.builtins, id)
+        ) {
+          if (
+            options.enableBuiltinNoExternalCheck &&
+            options.noExternal === true &&
+            // if both noExternal and external are true, noExternal will take the higher priority and bundle it.
+            // only if the id is explicitly listed in external, we will externalize it and skip this error.
+            (options.external === true || !options.external.includes(id))
+          ) {
+            let message = `Cannot bundle built-in module "${id}"`
+            if (importer) {
+              message += ` imported from "${path.relative(
+                process.cwd(),
+                importer,
+              )}"`
             }
+            message += `. Consider disabling environments.${this.environment.name}.noExternal or remove the built-in dependency.`
+            this.error(message)
+          }
 
-            return options.idOnly
-              ? id
-              : { id, external: true, moduleSideEffects: false }
-          } else {
-            if (!asSrc) {
-              debug?.(
-                `externalized node built-in "${id}" to empty module. ` +
-                  `(imported by: ${colors.white(colors.dim(importer))})`,
-              )
-            } else if (isProduction) {
-              this.warn(
-                `Module "${id}" has been externalized for browser compatibility, imported by "${importer}". ` +
-                  `See https://vite.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.`,
-              )
-            }
-            return isProduction
-              ? browserExternalId
-              : `${browserExternalId}:${id}`
+          return options.idOnly
+            ? id
+            : { id, external: true, moduleSideEffects: false }
+        } else if (
+          currentEnvironmentOptions.consumer === 'client' &&
+          isNodeLikeBuiltin(id)
+        ) {
```

Comment:
```suggestion
        } else if (
          currentEnvironmentOptions.consumer === 'server' &&
          isNodeLikeBuiltin(id)
        ) {
          if (
            options.noExternal === true &&
            // if both noExternal and external are true, noExternal will take the higher priority and bundle it.
            // only if the id is explicitly listed in external, we will externalize it and skip this error.
            (options.external === true || !options.external.includes(id))
          ) {
            let message = `Cannot bundle built-in module "${id}"`
            if (importer) {
              message += ` imported from "${path.relative(
                process.cwd(),
                importer,
              )}"`
            }
            message += `. Consider disabling environments.${this.environment.name}.noExternal or remove the built-in dependency.`
            this.error(message)
          }
        }if (
          currentEnvironmentOptions.consumer === 'client' &&
          isNodeLikeBuiltin(id)
        ) {
```
How about moving the "if options.enableBuiltinNoExternalCheck block" here from above and removing `options.enableBuiltinNoExternalCheck &&` and setting `builtin: []` by default for `ssr.target: 'webworker'`?

By setting `builtin: []`, I mean updating this line to `(consumer === 'server' && !isSsrTargetWebworkerEnvironment`.
https://github.com/vitejs/vite/pull/18584/files#diff-11e17761d4ecfee8f8fde15c6d79b7bc0260176396a30dfd8e6f6bbaf5af4745R888

This way if an environment sets `builtin` and have some node builtin modules not included in `builtin`, Vite will show a more friendly error (`Cannot bundle built-in module`) when `noExternal` is set.

Reply:
I've made the changes in regards to `enableBuiltinNoExternalCheck` that you suggested, it sounds sensible to me 🙂👍